u_pre_shared_key.go

```diff
@@ -0,0 +1,503 @@
+	b[5] = byte(identitiesLength)
+
+	// identities
+	offset := 6
+	for _, identity := range e.identities {
+		b[offset] = byte(len(identity.label) >> 8)
+		b[offset+1] = byte(len(identity.label))
+		offset += 2
+		copy(b[offset:], identity.label)
+		offset += len(identity.label)
+		b[offset] = byte(identity.obfuscatedTicketAge >> 24)
+		b[offset+1] = byte(identity.obfuscatedTicketAge >> 16)
+		b[offset+2] = byte(identity.obfuscatedTicketAge >> 8)
+		b[offset+3] = byte(identity.obfuscatedTicketAge)
+		offset += 4
+	}
+
+	// concatenate ClientHello and PreSharedKeyExtension
+	rawHelloSoFar := append(raw, b[:offset]...)
+	transcript := e.cipherSuite.hash.New()
+	transcript.Write(rawHelloSoFar)
+	e.binders = [][]byte{e.cipherSuite.finishedHash(e.binderKey, transcript)}
+
+	// binders length
+	bindersLength := 0
+	for _, binder := range e.binders {
+		bindersLength += len(binder) + 1 // binder length + binder
+	}
+	b[offset] = byte(bindersLength >> 8)
+	b[offset+1] = byte(bindersLength)
+	offset += 2
+
+	// binders
+	for _, binder := range e.binders {
+		b[offset] = byte(len(binder))
+		offset++
+		copy(b[offset:], binder)
+		offset += len(binder)
+	}
+
+	return e.Len(), io.EOF
+}
+
+func (e *UtlsPreSharedKeyExtension) preloadSession(uc *UConn) error {
+	// var sessionCache ClientSessionCache
+	// must set either e.Session or uc.config.ClientSessionCache
+	if e.SessionCacheOverride != nil {
+		uc.config.ClientSessionCache = e.SessionCacheOverride
+	}
+
+	// load Hello
+	hello := uc.HandshakeState.Hello.getPrivatePtr()
+	// try to use loadSession()
+	session, earlySecret, binderKey, err := uc.loadSession(hello)
+	if err != nil {
+		return err
+	}
+	if session != nil && session.version == VersionTLS13 && binderKey != nil {
+		e.identities = hello.pskIdentities
+		e.binders = hello.pskBinders
+		e.binderKey = binderKey
+		e.cipherSuite = cipherSuiteTLS13ByID(session.cipherSuite)
+		e.earlySecret = earlySecret
+		return nil
+	} else {
+		return errors.New("tls: session not compatible with TLS 1.3, PSK not possible")
+	}
+}
+
+// Binders must be called after ReadWithRawHello
+func (e *UtlsPreSharedKeyExtension) Binders() [][]byte {
+	return e.binders
+}
+
+// FakePreSharedKeyExtension is an extension used to send the PSK extension in the
+// ClientHello.
+//
+// However, it DOES NOT do any session resumption AND should not be used with a
+// real/valid PSK Identity.
+//
+// TODO: Only one of FakePreSharedKeyExtension and HardcodedPreSharedKeyExtension should
+// be kept, the other one should be just removed. We still need to learn more of the safety
+// of hardcoding both Identities and Binders without recalculating the latter.
+type FakePreSharedKeyExtension struct {
+	UnimplementedPreSharedKeyExtension
+
+	CipherSuite   uint16 `json:"cipher_suite"`   // this is used to compute the binder
+	SessionSecret []byte `json:"session_secret"` // this is used to compute the binder
+
+	Identities []PskIdentity `json:"identities"`
+	binders    [][]byte
+}
```

Comment:
We will remove the current implementation of `FakePreSharedKeyExtension` and use `HardcodedPreSharedKeyExtension` instead. The final naming is subjected to change. 

See https://github.com/refraction-networking/utls/issues/107#issuecomment-1679930782 for the discussion supporting this change.